Mathlib/MeasureTheory/Measure/SeparableMeasure.lean

```diff
@@ -0,0 +1,494 @@
+
+/-- A measure `μ` is separable if there exists a countable and measure-dense family of sets.
+
+The term "separable" is justified by the fact that the definition translates to the usual notion
+of separability in the metric space made by constant indicators equipped with the `L¹` norm. -/
+class IsSeparable (μ : Measure X) : Prop :=
+  exists_countable_measureDense : ∃ 𝒜, 𝒜.Countable ∧ μ.MeasureDense 𝒜
+
+/-- By definition, a separable measure admits a countable and measure-dense family of sets. -/
+theorem exists_countable_measureDense [IsSeparable μ] :
+    ∃ 𝒜, 𝒜.Countable ∧ μ.MeasureDense 𝒜 :=
+  IsSeparable.exists_countable_measureDense
+
+/-- If a measurable space is countably generated and equipped with a `σ`-finite measure, then the
+measure is separable. This is not an instance because it is used below to prove the more
+general case where `μ` is s-finite. -/
+theorem isSeparable_of_sigmaFinite [CountablyGenerated X] [SigmaFinite μ] :
+    IsSeparable μ where
+  exists_countable_measureDense := by
+    have h := countable_countableGeneratingSet (α := X)
+    have hgen := generateFrom_countableGeneratingSet (α := X)
+    let 𝒜 := (countableGeneratingSet X) ∪ {μ.toFiniteSpanningSetsIn.set n | n : ℕ}
+    have count_𝒜 : 𝒜.Countable :=
+      countable_union.2 ⟨h, countable_iff_exists_subset_range.2
+        ⟨μ.toFiniteSpanningSetsIn.set, fun _ hx ↦ hx⟩⟩
+    refine ⟨generateSetAlgebra 𝒜, countable_generateSetAlgebra count_𝒜,
+      measureDense_of_generateFrom_setAglebra_of_sigmaFinite isSetAlgebra_generateSetAlgebra
+      {
+        set := μ.toFiniteSpanningSetsIn.set
+        set_mem := fun n ↦ self_subset_generateSetAlgebra (𝒜 := 𝒜) <| Or.inr ⟨n, rfl⟩
+        finite := μ.toFiniteSpanningSetsIn.finite
+        spanning := μ.toFiniteSpanningSetsIn.spanning
+      }
+      (le_antisymm ?_ (generateFrom_le (fun s hs ↦ ?_)))⟩
+    · rw [← hgen]
+      exact generateFrom_mono <| le_trans self_subset_generateSetAlgebra <|
+        generateSetAlgebra_mono <| subset_union_left ..
+    · induction hs with
+      | base t t_mem =>
+        rcases t_mem with t_mem | ⟨n, rfl⟩
+        · exact hgen ▸ measurableSet_generateFrom t_mem
+        · exact μ.toFiniteSpanningSetsIn.set_mem n
+      | empty => exact MeasurableSet.empty
+      | compl t _ t_mem => exact MeasurableSet.compl t_mem
+      | union t u _ _ t_mem u_mem => exact MeasurableSet.union t_mem u_mem
+
+/-- If a measurable space is countably generated and equipped with an s-finite measure, then the
+measure is separable. -/
+instance instIsSeparableCountablyGeneratedSFinite [CountablyGenerated X] [SFinite μ] :
+    IsSeparable μ where
+  exists_countable_measureDense := by
+    have := isSeparable_of_sigmaFinite (μ := μ.restrict μ.sigmaFiniteSet)
+    rcases exists_countable_measureDense (μ := μ.restrict μ.sigmaFiniteSet) with ⟨𝒜, count_𝒜, h𝒜⟩
+    let ℬ := {s ∩ μ.sigmaFiniteSet | s ∈ 𝒜}
+    refine ⟨ℬ, count_𝒜.image (fun s ↦ s ∩ μ.sigmaFiniteSet), ?_⟩
+    constructor
```

Comment:
```suggestion
    refine ⟨ℬ, count_𝒜.image (fun s ↦ s ∩ μ.sigmaFiniteSet), ?_, ?_⟩
```
`refine` can already be used to split your goal in several subgoals when needed.